File: packages/python/anip-crypto/src/anip_crypto/verify.py
"""Verification helpers for ANIP signed artifacts."""

import hashlib

import jwt as pyjwt

from .canonicalize import canonicalize
from .keys import KeyManager


def verify_audit_entry_signature(
    key_manager: KeyManager, entry: dict, signature: str
) -> dict:
    """Verify an audit entry's signature using the audit public key.

    Returns the decoded JWT claims on success, raises on failure.
    """
    claims = pyjwt.decode(
        signature,
        key_manager.audit_public_key,
        algorithms=["ES256"],
    )
    canonical = canonicalize(entry, exclude={"signature", "id"})
    expected_hash = hashlib.sha256(canonical).hexdigest()
    if claims.get("audit_hash") != expected_hash:
        raise ValueError("Audit hash mismatch")
    return claims


def verify_manifest_signature(
    key_manager: KeyManager, manifest_bytes: bytes, signature: str
) -> None:
    """Verify a manifest's detached JWS signature using the delegation public key.

    *manifest_bytes* is the raw JSON body returned by ``GET /anip/manifest``.
    *signature* is the ``X-ANIP-Signature`` header value (detached JWS).

    Raises on verification failure.
    """
    key_manager.verify_jws_detached(signature, manifest_bytes)

File: packages/typescript/crypto/src/verify.ts
import { createHash } from "crypto";
import type { KeyManager } from "./keys.js";
import { canonicalize } from "./canonicalize.js";
import * as jose from "jose";

/**
 * Verify that an audit entry's JWT signature is valid and that the embedded
 * `audit_hash` matches the canonical hash of the entry data.
 *
 * Returns the decoded JWT claims on success; throws on any mismatch.
 */
export async function verifyAuditEntrySignature(
  km: KeyManager,
  entry: Record<string, unknown>,
  signature: string,
): Promise<jose.JWTPayload> {
  await km.ready();

  // Verify the JWT using the audit public key
  const { payload } = await jose.jwtVerify(
    signature,
    km.getAuditPublicKey(),
    { algorithms: ["ES256"] },
  );

  // Recompute the expected hash from entry data (excluding signature and id)
  const canonical = canonicalize(entry, new Set(["signature", "id"]));
  const expectedHash = createHash("sha256").update(canonical).digest("hex");

  if (payload.audit_hash !== expectedHash) {
    throw new Error(
      `Audit hash mismatch: expected ${expectedHash}, got ${payload.audit_hash as string}`,
    );
  }

  return payload;
}
